java/spark/v3.5/src/main/java/com/lancedb/lance/spark/source/SparkWrite.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.spark.sql.execution.arrow.ArrowWriter;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.ArrowUtils;
+
+/**
+ * Spark write.
+ */
+public class SparkWrite implements Write {
+  private final String datasetUri;
+  private final StructType  sparkSchema;
+  private final LogicalWriteInfo info;
+  
+  SparkWrite(String datasetUri, StructType sparkSchema, LogicalWriteInfo info) {
+    this.datasetUri = datasetUri;
+    this.sparkSchema = sparkSchema;
+    this.info = info;
+  }
+
+  @Override
+  public BatchWrite toBatch() {
+    return new BatchAppend();
+  }
+
+  @Override
+  public StreamingWrite toStreaming() {
+    throw new UnsupportedOperationException();
+  }
+
+  private WriterFactory createWriterFactory() {
+    return new WriterFactory(datasetUri, sparkSchema);
+  }
+
+  private class BatchAppend extends BaseBatchWrite {
+    @Override
+    public void commit(WriterCommitMessage[] messages) {
+      List<FragmentMetadata> fragments = Arrays.stream(messages)
+          .map(m -> (TaskCommit) m)
+          .map(TaskCommit::getFragments)
+          .flatMap(List::stream)
+          .collect(Collectors.toList());
+      FragmentOperation.Append appendOp = new FragmentOperation.Append(fragments);
+      try (BufferAllocator allocator = new RootAllocator()) {
+        Dataset.commit(allocator, datasetUri, appendOp, Optional.of(1));
+      }
+    }
+  }
+
+  private abstract class BaseBatchWrite implements BatchWrite {
+    @Override
+    public DataWriterFactory createBatchWriterFactory(PhysicalWriteInfo info) {
+      return createWriterFactory();
+    }
+
+    @Override
+    public boolean useCommitCoordinator() {
+      return false;
+    }
+
+    @Override
+    public void abort(WriterCommitMessage[] messages) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String toString() {
+      return String.format("LanceBatchWrite(datasetUri=%s)", datasetUri);
+    }
+  }
+
+  private static class WriterFactory implements DataWriterFactory {
+    private final String datasetUri;
+    private final StructType sparkSchema;
+
+    protected WriterFactory(String datasetUri, StructType sparkSchema) {
+      // Everything passed to writer factory should be serializable
+      this.datasetUri = datasetUri;
+      this.sparkSchema = sparkSchema;
+    }
+
+    @Override
+    public DataWriter<InternalRow> createWriter(int partitionId, long taskId) {
+      return new UnpartitionedDataWriter(datasetUri, sparkSchema);
+    }
+  }
+
+  private static class UnpartitionedDataWriter implements DataWriter<InternalRow> {
+    private final String datasetUri;
+    private final BufferAllocator allocator;
+    private final VectorSchemaRoot root;
+    private final ArrowWriter writer;
+    
+    private UnpartitionedDataWriter(String datasetUri, StructType sparkSchema) {
+      // TODO(lu) add Lance Spark configuration of maxRowsPerFragment?
```

Comment:
@eddyxu the java approach is write to VectorSchemaRoot, unload from vectorSchemaRoot to c.ArrowArray, and pass to Lance to write to a Lance Fragment.
If i do small batch write could result in a large amount of small Lance Fragment files. If i do large batch write, could result in high memory consumption. Haven't found an existing good Arrow supported approach that Spark can keep write batches to VectorSchemaRoot and i can keep unload batches and turn to c.ArrowArray and keep appending to same fragment

Reply:
Will improve this in following PR